AK/StringFloatingPointConversions.cpp

```diff
@@ -0,0 +1,2279 @@
+/*
+ * Copyright (c) 2022, Dan Klishch <danilklishch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/BitCast.h>
+#include <AK/BuiltinWrappers.h>
+#include <AK/StringFloatingPointConversions.h>
+
+namespace AK {
+
+// This entire algorithm is an implementation of the paper: Ryu: Fast Float-to-String Conversion
+// by Ulf Adams, available at https://dl.acm.org/doi/pdf/10.1145/3192366.3192369 and an implemenetation
+// at https://github.com/ulfjack/ryu.
+//
+// Currently implemented only for doubles.
+
+/* Step 0. Precompute lookup tables for the given floating point type.
+The lookup table was generated using the following Python script:
+
+from math import *
+
+def minmax_euclid(a, b, M):
+    assert gcd(a, b) == 1
+    assert a < b
+
+    if M >= b:
+        return (0, b - 1)
+
+    x, y = a, b
+    s, t, u, v = 1, 0, 0, 1
+    while True:
+        assert x * s + y * t == a
+        assert x * u + y * v == b
+
+        while b >= a:
+            b, u, v = b - a, u - s, v - t
+            if -u >= M:
+                return (a, b)
+        if b == 0:
+            return (1, b - 1)
+
+        while a >= b:
+            olda = a
+            a, s, t = a - b, s - u, t - v
+            if s >= M:
+                return (a, b)
+        if a == 0:
+            return (1, b - 1)
+
+maxw = 2 ** 55 - 2
+
+for e2 in range(-1076, 0):
+    q = max(0, floor(-e2 * log(5, 10)) - 1)
+    a = 5 ** (-e2 - q)
+    b = 2 ** q
+
+    [mn, mx] = minmax_euclid(a % b, b, maxw)
+
+    if mn < maxw:
+        k = 0
+    else:
+        k = floor(log2(mn / maxw))
+    coefficient = 5 ** (-e2 - q) // 2 ** k
+
+    print(f"\t{{{q}, {k}, {{{coefficient & ((1 << 64) - 1)}ull, {coefficient >> 64}ull}}}},")
+
+for e2 in range(0, 970):
+    q = max(0, floor(e2 * log(2, 10)) - 1)
+    a = 2 ** (e2 - q)
+    b = 5 ** q
+
+    [mn, mx] = minmax_euclid(a % b, b, maxw)
+
+    k = ceil(log2(maxw * a * b / (b - mx)))
+    coefficient = 2 ** k // 5 ** q + 1
+
+    print(f"\t{{{q}, {k}, {{{coefficient & ((1 << 64) - 1)}ull, {coefficient >> 64}ull}}}},")
+
+*/
+// FIXME: Should be possible to compress from the current (48Kb) size to the one
+//        mentioned in the article (~9.5Kb).
+struct {
+    i32 q;
+    i32 k;
+    u64 multiplier[2];
+} DOUBLE_LOOKUP_TABLE[] = {
```

Comment:
constants still use snake case in our coding style,
same in the implementation functions